discovery/mdns.md

```diff
@@ -0,0 +1,93 @@
+# Multicast DNS
+Author: Richard Schneider (makaretu@gmail.com)
+
+
+## Overview
+
+The goal is to allow peers to discover each other when on the same local network with zero configuration. 
+MDNS uses a multicast system of DNS records; this allows all peers on the local network to see all query responses.
+
+Conceptually, it is very simple.  When a peer starts (or detects a network change), it sends a query for all peers. 
+As responses come in, the peer adds the other peers information into is local database of peers.
+
+## Definitions
+
+`service-name` is the DNS-SD service name for all peers. It is defined as `_p2p._udp.local`.
+
+`host-name` is the fully qualified name of the peer.  It is derived from the peer's name and `p2p.local`.
+
+`peer-name` is the case-insenstive ID of the peer and less than 64 characters.  It is normally is the base-32 encoding of peer's ID.

```

Comment:
This limits PeerID to 320bits which might become a problem if ever we need to switch to 512bit hash function (just mentioning for you to consider).

Reply:
This is similar problem we have with using CIDs in subdomains.

One of proposed workarounds in https://github.com/ipfs/in-web-browsers/issues/89#issue-341357014 is to "Split at 63rd character"

Reply:
Good idea. I'll make the changes.

Reply:
It might be useful to create a separate doc for DNS encoding of CIDs. As it is already present in two places.

Reply:
Is there a technical reason to limit the peer id size?

Reply:
@vyzo  IIUC we need to comply with length limits from  [RFC 1034](https://tools.ietf.org/html/rfc1034#page-7):

> Each node has a label, which is zero to 63 octets in length.  [#](https://tools.ietf.org/html/rfc1034#page-7)

> To simplify implementations, the total number of octets that represent a
domain name (i.e., the sum of all label octets and label lengths) is
limited to 255. [#](https://tools.ietf.org/html/rfc1034#page-8)

CIDv1(sha-256) in Base32 fits, but hashes with longer outputs can't be used as a single label and need to be split.

Reply:
@vasco-santos The peer-name is used as a DNS label.  A DNS label must be case insensitive, less than 63 characters and US-ASCII.